tests/core/test_lightning_optimizer.py

```diff
@@ -0,0 +1,192 @@
+    """
+    class TestModel(BoringModel):
+
+        def training_step(self, batch, batch_idx, optimizer_idx=None):
+            (opt_1, opt_2) = self.optimizers()
+            assert isinstance(opt_1, LightningOptimizer)
+            assert isinstance(opt_2, LightningOptimizer)
+
+            output = self.layer(batch)
+            loss_1 = self.loss(batch, output)
+            self.manual_backward(loss_1, opt_1)
+            opt_1.step(idx="1")
+
+            def closure():
+                output = self.layer(batch)
+                loss_2 = self.loss(batch, output)
+                self.manual_backward(loss_2, opt_2)
+            opt_2.step(closure=closure, idx="2")
+
+        def configure_optimizers(self):
+            optimizer_1 = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+            optimizer_2 = torch.optim.Adam(self.layer.parameters(), lr=0.1)
+            optimizer_1 = LightningOptimizer(optimizer_1, 4)
+
+            lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer_1, step_size=1)
+            return [optimizer_1, optimizer_2], [lr_scheduler]
+
+    model = TestModel()
+    model.training_step_end = None
+    model.training_epoch_end = None
+    trainer = Trainer(
+        default_root_dir=os.getcwd(),
+        limit_train_batches=8,
+        limit_val_batches=1,
+        max_epochs=1,
+        weights_summary=None,
+        automatic_optimization=False,
+        enable_pl_optimizer=True)
+    trainer.fit(model)
+
+    assert len(mock_sgd_step.mock_calls) == 2
+    assert len(mock_adam_step.mock_calls) == 8
+
+
+@patch("torch.optim.Adam.step")
+@patch("torch.optim.SGD.step")
+def test_lightning_optimizer_manual_optimization_and_accumulated_gradients(mock_sgd_step, mock_adam_step, tmpdir):
+    """
+    Test that the user can use our LightningOptimizer. Not recommended.
+    """
+    class TestModel(BoringModel):
+
+        def training_step(self, batch, batch_idx, optimizer_idx=None):
+            (opt_1, opt_2) = self.optimizers()
+            assert isinstance(opt_1, LightningOptimizer)
+            assert isinstance(opt_2, LightningOptimizer)
+
+            output = self.layer(batch)
+            loss_1 = self.loss(batch, output)
+            self.manual_backward(loss_1, opt_1)
+            opt_1.step(idx="1")
+
+            def closure():
+                output = self.layer(batch)
+                loss_2 = self.loss(batch, output)
+                self.manual_backward(loss_2, opt_2)
+            opt_2.step(closure=closure, idx="2")
+
+        def configure_optimizers(self):
+            optimizer_1 = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+            optimizer_2 = torch.optim.Adam(self.layer.parameters(), lr=0.1)
+            optimizer_1 = LightningOptimizer(optimizer_1, 4)
+
+            lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer_1, step_size=1)
+            return [optimizer_1, optimizer_2], [lr_scheduler]
+
+    model = TestModel()
+    model.training_step_end = None
```

Comment:
rather define it the class, this is easily overlooked

Reply:
Not sure to understand what you mean there. It is the current way to do. We override the model function.